Add postpublish hook

mmaclach opened the pull request:
Add postpublish hook used to modify files after publishing, or to kick off other processes. One concrete example is to update lock files, which currently don't get updated (using npm)--see #482.

Fixes #488 
Fixes #482

ecraig12345 commented on src/publish/publishToRegistry.ts:
```diff
@@ -43,11 +43,15 @@ export async function publishToRegistry(originalBumpInfo: BumpInfo, options: Bea
-      const packageInfo = bumpInfo.packageInfos[pkg];
+
+  // finally pass through doing the actual npm publish command
+  for (const pkg of packagesToPublish) {
```
I'm not sure if the change to order/grouping of operations here is right (but I can kind of see arguments for either ordering). @kenotron or @xugao what do you think?

Before:
prepublish all packages, publish all packages, (new) postpublish all packages

After:
for each package, prepublish/publish/postpublish

mmaclach replied:
Per the comment below I will go ahead and revert the grouping change.

kenotron commented:
There should be some description of the use case for this feature, and also the approach in theory. For example, the comment was called out in the PR that there's now a new kind of grouping. In general, the prepub & maybe the postpub hook would be executed in phases. If a package level prepub and postpub steps are needed, I think npm lifecycle scripts are what is needed.

kenotron reviewed (changes requested):
Please add a description of what the use case is. Preferably in an issue as a feature request?

mmaclach commented:
Hi @kenotron 

I've added a feature request here: https://github.com/microsoft/beachball/issues/488

For the approach, I followed the example of the existing one and moved the hooks inside the main publish loop. Re: your comment of keeping the old grouping, I will go ahead and revert this grouping change.

ecraig12345 commented on src/types/BeachballOptions.ts:
```diff
@@ -73,6 +73,12 @@ export interface RepoOptions {
+
+    /**
+     * Runs for each package after the publish command.
+     *
```
```suggestion
```

ecraig12345 reviewed (approved):
Thanks for the updates!

ecraig12345 commented:
@mmaclach In the future it's preferable to include a description in the PR description field itself, and just link to the existing issue (no need to create a new one). I updated it this time and cleared the request changes status.

ecraig12345 commented:
> If a package level prepub and postpub steps are needed, I think npm lifecycle scripts are what is needed.

@kenotron I think the current approach is reasonable since it's already what we do for prepublish (you added it in #319).

ecraig12345 commented on src/types/BeachballOptions.ts:
```diff
@@ -73,6 +73,11 @@ export interface RepoOptions {
     prepublish?: (packagePath: string, name: string, version: string) => void | Promise<void>;
+
+    /**
+     * Runs for each package after the publish command.
```
```suggestion
     * Runs for each package after the publish command.
     * Any file changes made in this step will **not** be committed automatically.
```

mmaclach commented:
> @mmaclach In the future it's preferable to include a description in the PR description field itself, and just link to the existing issue (no need to create a new one). I updated it this time and cleared the request changes status.

sounds good, thanks @ecraig12345

kenotron commented:
this looks good to me as well
